File: 1367.linked-list-in-binary-tree.java
#
# @lc app=leetcode id=1367 lang=java
#
# [1367] Linked List in Binary Tree
#
# @lc code=start
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
/**
* Definition for a binary tree node.
* public class TreeNode {
*     int val;
*     TreeNode left;
*     TreeNode right;
*     TreeNode() {}
*     TreeNode(int val) { this.val = val; }
*     TreeNode(int val, TreeNode left, TreeNode right) {
*         this.val = val;
*         this.left = left;
*         this.right = right;
*     }
* }
*/
class Solution {
    public boolean isSubPath(ListNode head, TreeNode root) {
        if (root == null) return false;
        // Check match starting at current node, left, or right
        return checkPath(head, root) || isSubPath(head, root.left) || isSubPath(head, root.right);
    }
    // Helper to check if path from root matches the linked list
    private boolean checkPath(ListNode head, TreeNode node) {
        if (head == null) return true;
        if (node == null) return false;
        if (head.val != node.val) return false;
        return checkPath(head.next, node.left) || checkPath(head.next, node.right);
    }
}
# @lc code=end

File: 1367.linked-list-in-binary-tree.go
#
# @lc app=leetcode id=1367 lang=golang
#
# [1367] Linked List in Binary Tree
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubPath(head *ListNode, root *TreeNode) bool {
    var check func(*TreeNode, *ListNode) bool
    check = func(node *TreeNode, lis *ListNode) bool {
        if lis == nil {
            return true
        }
        if node == nil || node.Val != lis.Val {
            return false
        }
        return check(node.Left, lis.Next) || check(node.Right, lis.Next)
    }

    var dfs func(*TreeNode) bool
    dfs = func(node *TreeNode) bool {
        if node == nil {
            return false
        }
        if node.Val == head.Val && check(node, head) {
            return true
        }
        return dfs(node.Left) || dfs(node.Right)
    }

    return dfs(root)
}
# @lc code=end